s/core.go

```diff
@@ -0,0 +1,269 @@
+package s
+
+import (
+	"context"
+	"strings"
+
+	"github.com/capatazlib/go-capataz/c"
+)
+
+// WithOrder specifies the start/stop order of a supervisor's children
+func WithOrder(o Order) Opt {
+	return func(spec *Spec) {
+		spec.order = o
+	}
+}
+
+// WithStrategy specifies how children get restarted when one of them fails
+func WithStrategy(s Strategy) Opt {
+	return func(spec *Spec) {
+		spec.strategy = s
+	}
+}
+
+// WithNotifier specifies a callback that gets called whenever the supervision
+// system reports an Event
+func WithNotifier(en EventNotifier) Opt {
+	return func(spec *Spec) {
+		spec.eventNotifier = en
+	}
+}
+
+// WithChildren specifies a list of child Spec that will get started when the
+// supervisor starts
+func WithChildren(children ...c.Spec) Opt {
+	return func(spec *Spec) {
+		spec.children = append(spec.children, children...)
+	}
+}
+
+// WithSubtree specifies a supervisor sub-tree. Is intended to be used when
+// composing sub-systems in a supervision tree.
+func WithSubtree(subtree Spec, copts ...c.Opt) Opt {
+	return func(spec *Spec) {
+		cspec := spec.Subtree(subtree, copts...)
+		WithChildren(cspec)(spec)
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Supervisor (dynamic tree) functionality
+
+// handleChildResult returns a callback function that gets called by a
+// supervised child whenever it finishes it main execution function.
+func (sup Supervisor) handleChildResult() func(string, error) {
+	// The function bellow gets called in the child goroutine
+	return func(childName string, err error) {
+		if err != nil {
+			// TODO report failed
+		} else {
+			// TODO report finished
+		}
+	}
+}
+
+// Stop is a synchronous procedure that halts the execution of the whole
+// supervision tree.
+func (sup Supervisor) Stop() error {
+	sup.cancel()
+	err := sup.wait()
+	sup.spec.getEventNotifier().ProcessStopped(sup.runtimeName, err)
+	return err
+}
+
+// Wait blocks the execution of the current goroutine until the Supervisor
+// finishes it execution.
+func (sup Supervisor) Wait() error {
+	return sup.wait()
+}
+
+// Name returns the name of the Spec used to start this Supervisor
+func (sup Supervisor) Name() string {
+	return sup.spec.Name()
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Spec (static tree) functionality
+
+// emptyEventNotifier is an utility function that works as a default value
+// whenever an EventNotifier is not specified on the Supervisor Spec
+func emptyEventNotifier(_ Event) {}
+
+// getEventNotifier returns the configured EventNotifier or emptyEventNotifier
+// (if none is given via WithEventNotifier)
+func (spec Spec) getEventNotifier() EventNotifier {
+	if spec.eventNotifier == nil {
```

Comment:
I would not do it this way. Why not just put the `emptyEventNotifier` in the the `New()` factory function? If you really got crazy, you already have accessor methods, so just create a Spec interface and make the struct private. That way you are guaranteed to have a sanely constructed spec.

Reply:
Tried out to code this approach, and is way more code, dealing with a lower case `supervisorSpec` (struct) and a public `SupervisorSpec` (interface). This interface is only going to be used by supervisorSpec, so creating the interface in the first place feels overkill. I'm going to change it so that we have the `emptyEventNotifier` on the default, but I will keep it as is as there is less LOC

Reply:
Lots more code is the java way and it is very rarely the best way. Keep it simple. Thanks for considering it.